Manual rule run from rule details and rules table (#9327)

e40pud opened the pull request:
## Summary

Main ticket https://github.com/elastic/security-team/issues/9327

With this changes we introduce the way to schedule rule run manually. There are two ways to do that in UI:
1. Via "All actions" button on rules management page
2. Via "All actions" button on rule's details page

**NOTES**:
1. To be able to test these changes, you need to enable feature flag `manualRuleRunEnabled` first
2. Bulk action will be part of a separate ticket/PR

**RECORDING**:

https://github.com/elastic/kibana/assets/2700761/d49bad53-026e-49c2-aeea-481203260b23

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
    - [x] https://github.com/elastic/security-docs/issues/5264
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
    - [x] [Cypress RM (100 ESS & 100 Serverless)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6263)
    - [x] [Cypress DE (100 ESS & 100 Serverless)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6280)
    - [x] [Integration Rule Gaps (100 ESS & 100 Serverless)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6257)

e40pud committed:
Manual rule run from rule details and rules table (#9327)

e40pud committed:
Fix types

e40pud commented:
/ci

e40pud committed:
Fix tests

e40pud commented:
/ci

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6174](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6174)
[✅] Security Solution Rule Management - Cypress: 50/50 tests passed.
[✅] Security Solution Detection Engine - Cypress: 50/50 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 50/50 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 24/50 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

e40pud committed:
Integration tests

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

ymao1 reviewed (approved):
Response Ops changes LGTM. Code review only of single affected file.

e40pud commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/manual_rule_run.ts:
```diff
@@ -0,0 +1,285 @@
+        );
+      });
+
+      it('should return bad request error when some of the rules are disabledx', async () => {
```
```suggestion
      it('should return bad request error when some of the rules are disabled', async () => {
```

e40pud committed:
Update x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/manual_rule_run.ts

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

dplumlee commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/components/manual_rule_run/index.tsx:
```diff
@@ -0,0 +1,124 @@
+
+  const [startDate, setStartDate] = useState(moment().subtract(1, 'd'));
+  const [endDate, setEndDate] = useState(moment());
+  const isInvalid = startDate.isAfter(endDate);
```
It looks like we can submit but the API throws an error if `startDate` and `endDate` are equal. Super minor edge case but if it's not too hard, we should match this verification.
<img width="326" alt="Screenshot 2024-06-03 at 12 31 01 PM" src="https://github.com/elastic/kibana/assets/56367316/7eb56af2-010a-486f-9093-84f27723a243">

e40pud replied:
Good point! Fixed

dplumlee commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/components/manual_rule_run/index.tsx:
```diff
@@ -0,0 +1,124 @@
+              />
+            }
+            endDateControl={
+              <EuiDatePicker
+                aria-label="End date range"
+                selected={endDate}
+                onChange={(date) => date && setEndDate(date)}
+                startDate={startDate}
+                endDate={endDate}
+                showTimeSelect={true}
```
Another UI nit: I can select dates way out into the future even though the API won't select any dates greater than `now`, have you messed around at all with the `minDate`/`maxDate` stuff that `EuiDatePicker` has at all?
<img width="505" alt="Screenshot 2024-06-03 at 12 42 04 PM" src="https://github.com/elastic/kibana/assets/56367316/d39720d0-d23e-461d-9fd4-fe873048cc57">

e40pud replied:
Tried max date, but it works only for the day selection. Anyway fixed it by showing error when user selects time range which is in the future.

dplumlee reviewed (approved):
This is so awesome to finally see in action, thanks @e40pud! Rule management changes look good to me, I just left some comments about the UI

e40pud committed:
Fixing serverless tests

e40pud committed:
Review feedback - error handling

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6210](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6210)
[❌] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/ess.config.ts: 74/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

MadameSheema commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_gaps/manual_rule_run.cy.ts:
```diff
@@ -0,0 +1,73 @@
+import { createRule } from '../../../../tasks/api_calls/rules';
+import { login } from '../../../../tasks/login';
+
+describe('Manual rule run', { tags: ['@ess', '@serverless'] }, () => {
```
I see that this test needs a FF to be able to pass. Currently FF are not supported on MKI environments, so this test and all the ones that might be affected by it (including API ones), should be skipped from MKI environments.

e40pud replied:
done

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6208](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6208)
[✅] Security Solution Rule Management - Cypress: 100/100 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6209](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6209)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 62/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

e40pud committed:
Review feedback: skip tests in MKI environments

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6217](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6217)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

MadameSheema reviewed (approved):
sec-eng-prod changes LGTM!

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6216](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6216)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 61/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/components/manual_rule_run/index.tsx:
```diff
@@ -0,0 +1,137 @@
+const ManualRuleRunModalComponent = ({ onCancel, onConfirm }: ManualRuleRunModalProps) => {
+  const modalTitleId = useGeneratedHtmlId();
+
+  const now = useMemo(() => moment(), []);
```
useMemo feels redundant here, as on every render now should change and reflect current time instead of memoized

e40pud replied:
done

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/components/manual_rule_run/index.tsx:
```diff
@@ -0,0 +1,137 @@
+
+  const now = useMemo(() => moment(), []);
+
+  const [startDate, setStartDate] = useState(now.clone().subtract(1, 'd'));
```
why start date is one day in past? Can we add comment here to explain it

e40pud replied:
done

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/components/manual_rule_run/index.test.tsx:
```diff
@@ -0,0 +1,54 @@
+ */
+
+import React from 'react';
+import { mount } from 'enzyme';
+import { render } from '@testing-library/react';
```
why mixing 2 testing libraries in one file? In new tests we try to stick to one choice: testing-library

e40pud replied:
done

vitaliidm commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/serverless.config.ts:
```diff
@@ -0,0 +1,25 @@
+      '/testing_regex*/',
+    ])}`, // See tests within the file "ignore_fields.ts" which use these values in "alertIgnoreFields"
+    `--xpack.securitySolution.enableExperimental=${JSON.stringify([
+      'bulkCustomHighlightedFieldsEnabled',
+      'alertSuppressionForEsqlRuleEnabled',
```
do we need these feature flags for rule gaps tests?

e40pud replied:
nope, deleted it

vitaliidm commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/manual_rule_run.ts:
```diff
@@ -0,0 +1,317 @@
+        });
+
+        expect(results.length).toBe(1);
+        expect((results[0] as any).start).toEqual(startDate.toISOString());
+        expect((results[0] as any).schedule).toEqual(
```
can we avoid using `any`?

e40pud replied:
done

vitaliidm commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_gaps/manual_rule_run.cy.ts:
```diff
@@ -0,0 +1,73 @@
+import { createRule } from '../../../../tasks/api_calls/rules';
+import { login } from '../../../../tasks/login';
+
+describe('Manual rule run', { tags: ['@ess', '@serverless', '@skipInServerlessMKI'] }, () => {
```
can we add comment here why it is skipped in MKI?

e40pud replied:
done

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/use_rules_table_actions.tsx:
```diff
@@ -109,6 +117,29 @@ export const useRulesTableActions = ({
+                timeRange: modalManualRuleRunConfirmationResult,
+              });
+            },
+            enabled: (rule: Rule) => rule.enabled,
```
you are checking for user permissions on details page: `x-pack/plugins/security_solution/public/detections/components/rules/rule_actions_overflow/index.tsx`

but not here. Should user permissions check also be added here?

e40pud replied:
It is already checked in [the parent component](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/use_columns.tsx#L346), that's why no permissions checks are done inside this hook

vitaliidm commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/index.ts:
```diff
@@ -0,0 +1,14 @@
+import { FtrProviderContext } from '../../../../../ftr_provider_context';
+
+export default function ({ loadTestFile }: FtrProviderContext) {
+  describe('Rule execution logic API', function () {
```
should it be `Rule gaps API`?

e40pud replied:
done

vitaliidm commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/manual_rule_run.ts:
```diff
@@ -0,0 +1,317 @@
+  const log = getService('log');
+  const es = getService('es');
+
+  describe('@ess @serverless @skipInServerlessMKI manual_rule_run', () => {
```
can we add comment why tests skipped in MKI?

e40pud replied:
done

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/components/manual_rule_run/index.tsx:
```diff
@@ -0,0 +1,137 @@
+  const [endDate, setEndDate] = useState(now.clone());
+
+  const isEndDateInFuture = endDate.isAfter(now);
+  const isInvalid = isEndDateInFuture || startDate.isSameOrAfter(endDate);
```
can we add validation for max date in past as on API side?

<img width="2300" alt="Screenshot 2024-06-06 at 12 37 16" src="https://github.com/elastic/kibana/assets/92328789/4fc70d8b-2121-41cd-a667-ccd4a088829c">

e40pud replied:
good catch, I missed this error

vitaliidm reviewed (commented):
@e40pud 

Thank you working on this feature, this looks great.

I played around and did not find any major issue. Left few minor comments.


Also, it is not related to this PR, but I have noticed that auto-refresh option on scheduled rule runs gets reset every time I leave that tab


https://github.com/elastic/kibana/assets/92328789/757a4a47-46e5-404e-a240-3ff2b0a25176

Is this known issue or intended behaviour? If not, can we create a ticket to address it?

I also seen Cypress tests unfortunately not very stable:
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6216

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6228](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6228)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 62/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

e40pud commented:
> @e40pud
> 
> Thank you working on this feature, this looks great.
> 
> I played around and did not find any major issue. Left few minor comments.
> 
> Also, it is not related to this PR, but I have noticed that auto-refresh option on scheduled rule runs gets reset every time I leave that tab
> 
>  Screen.Recording.2024-06-06.at.12.49.02.mov 
> 
> Is this known issue or intended behaviour? If not, can we create a ticket to address it?
> 
> I also seen Cypress tests unfortunately not very stable: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6216

Thanks for the review! Created ticket https://github.com/elastic/security-team/issues/9690. As for the cypress tests, yes they are unstable on server less and I will need to have a look into that.

e40pud committed:
Review feedback

- use testing-library only
- handle max 90 days lookback range
- no nested ternary operators
- additional tests

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/components/manual_rule_run/index.tsx:
```diff
@@ -0,0 +1,137 @@
+  const isInvalid = isEndDateInFuture || startDate.isSameOrAfter(endDate);
+  const errorMessage = useMemo(
+    () =>
+      isInvalid
```
done

e40pud committed:
Add comments about skipping tests on MKI

e40pud commented:
elasticmachine merge upstream

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

e40pud committed:
Fix cypress tests

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6245](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6245)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

rylnd commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/translations.ts:
```diff
@@ -125,3 +125,20 @@ export const BACKFILL_TABLE_SUBTITLE = i18n.translate(
+    {
+      values: { numRules },
+      defaultMessage:
+        'Successfully scheduled backfill for {numRules} {numRules, plural, =1 {rule} other {rules}}',
```
This syntax looks slightly different than the translation above with `maxDaysLookback`. Are they still equivalent?

e40pud replied:
Yeah they are equivalent at the end, but I will update to use the same style. The special # character above will be replaced by the count variable when we resolve the message.

rylnd reviewed (approved):
Excellent work here! Very well-organized and easy to follow.

I had a few nits, and a question about translation syntax, but nothing that should block this being merged.

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_gaps/api/api.ts:
```diff
@@ -6,11 +6,44 @@
+ *
+ * @param ruleIds Rule ids
+ * @param timeRange Time range
+ * @param signal to cancel request
```
No, it should be removed. Thanks for catching this!

e40pud committed:
Apply suggestions from code review

Co-authored-by: Ryland Herrick <ryalnd@gmail.com>

e40pud committed:
Review feedback

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6243](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6243)
[✅] Security Solution Rule Management - Cypress: 100/100 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6244](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6244)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 66/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

e40pud committed:
Fix cypress tests

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6257](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6257)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6255](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6255)
[❌] Security Solution Rule Management - Cypress: 74/100 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6256](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6256)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 69/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6263](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6263)
[✅] Security Solution Rule Management - Cypress: 100/100 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

kibanamachine committed:
Merge branch 'main' into security/feature/9327-manual-rule-run

vitaliidm reviewed (approved):
Thank you for addressing feedback.

Looks like Cypress tests still a bit unstable

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6280](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6280)
[❌] Security Solution Detection Engine - Cypress: 74/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 99/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F184500%2Fhead)
